cirq/sim/clifford_simulator.py

```diff
@@ -0,0 +1,749 @@
+                elif self.xs[i,k] & self.zs[i,k]:
+                    stabilizer += "Y%d" % k
+            
+            stabilizers.append(stabilizer)
+            
+        return stabilizers
+                
+    def _measure(self,q):
+        ''' Performs a projective measurement on the q'th qubit. 
+
+        Returns: the result (0 or 1) of the measurement.
+        '''
+        is_commuting = True
+        for i in range(self.n,2*self.n):
+            if self.xs[i,q]:
+                p = i
+                is_commuting = False
+                break
+                
+                
+        if is_commuting:
+            self.xs[2*self.n,:] = False
+            self.zs[2*self.n,:] = False
+            self.rs[2*self.n] = False
+            
+            for i in range(self.n):
+                if self.xs[i,q]:
+                    self._rowsum(2*self.n,self.n+i)
+            return int(self.rs[2*self.n])
+        
+        else:
+            for i in range(2*self.n):
+                if i != p and self.xs[i,q]:
+                    self._rowsum(i,p)
+
+            self.xs[p-self.n,:] = self.xs[p,:]
+            self.zs[p-self.n,:] = self.zs[p,:]
+            self.rs[p-self.n] = self.rs[p]
+
+            self.xs[p,:] = False
+            self.zs[p,:] = False
+
+            self.zs[p,q] = True
+
+            self.rs[p] = bool(np.random.randint(2))
+
+            return int(self.rs[p])
+
+
+class CH_Form():
+    '''
+    A representation of stabilizer states using the CH form,
+
+        |psi> = omega U_C U_H |s>
+
+    This representation keeps track of overall phase.
+
+    See Bravyi et al, 2016 for details.
+    '''
+    def __init__(self,num_qubits,initial_state=0):
+        self.n = num_qubits
+        
+        # The state is represented by a set of binary matrices and vectors.
+        # See Section IVa of Bravyi et al
+        self.G = np.eye(self.n,dtype=bool)
+        
+        self.F = np.eye(self.n,dtype=bool)
+        self.M = np.zeros((self.n,self.n),dtype=bool)
+        self.gamma = np.zeros(self.n,dtype=int)
+        
+        self.v = np.zeros(self.n,dtype=bool)
+        self.s = np.zeros(self.n,dtype=bool)
+        
+        self.omega = 1
+
+        def bits(s):
+            while s > 0:
+                yield s & 1
+                s >>= 1
+
+        # Apply X for every non-zero element of initial_state
+        for (i,val) in enumerate(bits(initial_state)):
+            if val:
+                self._X(self.n-i-1)
+
+    def copy(self):
+        copy = CH_Form(self.n)
+
+        copy.G = self.G.copy()
+        copy.F = self.F.copy()
+        copy.M = self.M.copy()
+        copy.gamma = self.gamma.copy()
+        copy.v = self.v.copy()
+        copy.s = self.s.copy()
+        copy.omega = self.omega
+
+        return copy
+    
+    def __repr__(self):
+        ''' Return the wavefunction representation of the state. '''
+        return cirq.dirac_notation(self.to_numpy())  
```

Comment:
The repr has to be an evaluable python expression that returns an instance equal to the original class.

(E.g. after modifying the constructor this would just be passing the current instance's attributes' reprs into the constructor)

Reply:
Done.